torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +588,167 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
+    auto checkSplitSizes = [this](std::vector<int> split_sizes, at::Tensor &tensor) {
+      if(split_sizes.size() == 0) {
+        if(tensor.size(0) % size_ != 0) {
+          throw std::runtime_error("Tensor's dim 0 does not divide equally across group size");
+        }
+      } else {
+        if(split_sizes.size() != size_) {
+          throw std::runtime_error("Number of tensor splits not equal to group size");
+        }
+        int sum = std::accumulate(split_sizes.begin(), split_sizes.end(), 0);
+        if(sum != tensor.size(0)) {
+          throw std::runtime_error("Split sizes doesn't match total dim 0 size");
+        }
+      }
+    };
+    checkSplitSizes(inputSplitSizes, inputTensors[0]);
+    checkSplitSizes(outputSplitSizes, outputTensors[0]);
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this, inputSplitSizes, outputSplitSizes](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        std::vector<int> send_lengths(size_);
+        std::vector<int> recv_lengths(size_);
+        std::vector<int> send_offsets(size_);
+        std::vector<int> recv_offsets(size_);
+        auto computeLengthsAndOffsets = [this](std::vector<int> split_sizes, at::Tensor &tensor, std::vector<int> &lengths, std::vector<int> &offsets) {
+          if(split_sizes.size() == 0) {
+            int split_size = tensor.numel() / size_;
+            for(int i = 0; i < size_; i++) {
+              lengths[i] = split_size;
+              offsets[i] = i * split_size;
+            }
+          } else {
+            int split_size = tensor.numel() / tensor.size(0);
+            int offset = 0;
+            for(int i = 0; i < size_; i++) {
+              int length = split_size * split_sizes[i];
+              lengths[i] = length;
+              offsets[i] = offset;
+              offset += length;
+            }
+          }
+        };
+        computeLengthsAndOffsets(inputSplitSizes, srcdata, send_lengths, send_offsets);
+        computeLengthsAndOffsets(outputSplitSizes, dstdata, recv_lengths, recv_offsets);
+        c10::DeviceGuard guard(srcdata.device());
+        std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+        MPI_CHECK(MPI_Alltoallv(
+              srcdata.data_ptr(),
+              send_lengths.data(),
+              send_offsets.data(),
+              mpiDatatype.at(srcdata.scalar_type()),
+              dstdata.data_ptr(),
+              recv_lengths.data(),
+              recv_offsets.data(),
+              mpiDatatype.at(dstdata.scalar_type()),
+              pgComm_));
+      };
+    auto entry = std::unique_ptr<WorkEntry>(
+        new WorkEntry(&inputTensors, &outputTensors, std::move(runFunc)));
+    return enqueue(std::move(entry));
+  }
+}
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::alltoall(
+    std::vector<std::vector<at::Tensor>>& outputTensors,
+    std::vector<std::vector<at::Tensor>>& inputTensors,
```

Comment:
do we need to flatten inputs/outputs?

Reply:
Sorry, didn't get what are you referring to..